### PR TITLE
[PyTorch] New expand_inplace API with MaybeOwned<Tensor> and no unary tuples

### PR DESCRIPTION
**Pull request by swolchok**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55351 [PyTorch] Fix const correctness for resize native functions
* #55685 [PyTorch] Avoid double indirection in MaybeOwned's borrowed state
* #55684 [PyTorch] Change MaybeOwned tests to use intrusive_ptr and Tensor
* #55553 [PyTorch] Mark borrowed case as C10_LIKELY in MaybeOwned
* #55419 [PyTorch] Allow copy operations on MaybeOwned
* #55316 [PyTorch] Add & use inferExpandGeometry_dimvector
* #55247 [PyTorch][easy] Use sizes()[x] instead of size(x) in addr
* #55246 [PyTorch] Update expand_size API to match expand_inplace
* #55245 [PyTorch] Update expand_outplace API to match expand_inplace
* #55244 [PyTorch] Add MaybeOwned::operator*() &&
* #55180 [PyTorch] Use infer_size_dimvector in ExpandUtils
* **#55065 [PyTorch] New expand_inplace API with MaybeOwned<Tensor> and no unary tuples**

expand_inplace may give you the same Tensor(s) back, and it unnecessarily wrapped single-Tensor results in a tuple. Further diffs will deprecate and replace the rest of the similar APIs in ExpandUtils.

Differential Revision: [D27469297](https://our.internmc.facebook.com/intern/diff/D27469297/)

### CODE REVIEW

**Comment by swolchok** [comment]
not ready for review; MaybeOwned in the result makes it really easy to create a dangling pointer

File: aten/src/ATen/ExpandUtils.h [hunk 1]
```diff
@@ -45,33 +46,81 @@ inline void check_defined(std::initializer_list<std::reference_wrapper<const Ten
...
+// expand_inplace_v2 may borrow from to_expand; prevent calling with an
+// obvious temporary.
+inline c10::MaybeOwned<Tensor> expand_inplace_v2(const Tensor& tensor, Tensor&& to_expand) = delete;
 
-  return std::make_tuple(to_expand.expand(tensor.sizes(), /*implicit=*/true)); // see [expand implicit]
+
+C10_DEPRECATED_MESSAGE("expand_inplace is deprecated. Use expand_inplace_v2 instead.")
+inline std::tuple<Tensor> expand_inplace(const Tensor &tensor, const Tensor &to_expand) {
+  return std::make_tuple(*expand_inplace_v2(tensor, to_expand));
 }
```
Comment (ezyang) [inline aten/src/ATen/ExpandUtils.h:64]: IMO let's just YOLO remove it

**Review by ezyang — APPROVED** [review]
okey dokey

**Comment by ezyang** [comment]
Since this is kind of advanced C++ it might be helpful for end users to have some usage documentation explaining how to use it safely.

File: aten/src/ATen/ExpandUtils.h (hunk 1 above)
Comment (swolchok) [reply aten/src/ATen/ExpandUtils.h:64]: OK, but then we need feature:bc-breaking, right?

**Comment by swolchok** [comment]
adding module:bc-breaking to be conservative  because people technically could have called these even if they're not supposed to be external-facing

**Pull request closed without merging**